decisions/2024-02-28_file_sync.md

```diff
@@ -0,0 +1,91 @@
+
+![File Sync](./media/file_sync.png)
+
+_Pros:_
+
+- Keeps the main sync process simple/unchanged
+- Allows for targeted logic for file syncing (e.g rate limiting in future)
+- Allows for files to be downloaded where they're needed rather than needing to sync everywhere (via std sync process)
+
+_Cons:_
+
+- We could endup writing quite a bit of code to implement this and may have unexexpcted challenges with the implementation
+- Other sync protocols might provide benefits e.g. rsync for example uses a delta transfer algorithm which only sends the differences between files, rather than the whole file.
+
+
+### Option 2 - Base64 encode files and use std sync
+
+Any files needed to be synced could be base64 encoded and sent as part of the normal sync process. This would be a simiplish process, but would mean that sync would be blocked until all files are fully downloaded. Base64 encoding would also increase the size of the file by 33% (or more) which could be a problem for large files.
+
+This option is probably only viable if we built in some kind of sync record prioritisation, which would make the process more complex.
+
+_Pros:_
+
+- Simple to implement (if we ignore the sync record prioritisation bit)
+
+_Cons:_
+
+- Sync would be blocked until all files are fully downloaded
+- Base64 encoding would increase the size of the file by 33% (or more)
+
+### Option 3 - Use a third party file sync process
+
+A tool such as (https://syncthing.net/)[https://syncthing.net/] or even dropbox, could be setup on every installation to sync files. This would be a simple process, but would mean that we'd need to setup and maintain a third party tool on every installation. This would save us development effort on creating our own file sync process.
+
+_Pros:_
+
+- Not much to develop, but there's still complexities around what files are available, how do we communicate this to the client?
+- Syncthing is a peer to peer protocol, so concievably we could sync files between installations without needing to go via the central server. (unlikely in practice, but possible)
+
+_Cons:_
+
+- How would this work with Android? Can dropbox even be accessed from our code base?
+- We'd need to setup and maintain a third party tool on every installation, even if the time taken for this is acceptable, the risk of miss configurations, missed steps in configuration, etc is high.
+- We probably wouldn't have a lot of control over which files go where, which might cause problems with disk usage, espeically on smaller devices.
+
+
+### Option 4 - Only store files on central server and query central directly for files when needed
+
+Would be the simpliest solution, but wouldn't meet our offline first goals
+
+### Option 5 - Custom implementation of an existing sync protocol
+
+https://github.com/abusch/st-rust perhaps? It's explicitly marked as a `Toy` implementation though.
+If there's a good candidate someone knows of, we could consider this.
+
```

Comment:
```suggestion
### Option 6 - file_list and FTP server

Perhaps use the file_list idea to control relating records, then use [ftp](https://crates.io/crates/ftp) to retrieve from central server? 

Pros: 
- Maybe ftp will be reliable? 
- [ftp](https://crates.io/crates/ftp) seems pretty clean, but is only the client side...
- Presumably heaps of ftp server solutions out there. 

Cons:
- Have to run an ftp server on the central server
  - Complexity? Maintenance?
  - Security? An open port to an ftp server and whatever vulnerabilities it might have...
  - Someone with ftp credentials might be able to download all documents? What can be put in place to constrain this? I'm not familiar with ftp perhaps can configure "users" for sites that are constrained by what they should be able to get via "file_list".  

```

Reply:
@Chris-Petty Happy to include this, but not sure it's going to add any advantages over HTTP which we already have implemented...